backend/src/main/kotlin/org/loculus/backend/utils/EarliestReleaseDateFinder.kt

```diff
@@ -0,0 +1,56 @@
+package org.loculus.backend.utils
+
+import kotlinx.datetime.LocalDate
+import kotlinx.datetime.LocalDateTime
+import kotlinx.datetime.LocalTime
+import mu.KotlinLogging
+import org.loculus.backend.service.submission.RawProcessedData
+
+private val log = KotlinLogging.logger { }
+
+/**
+ * Calculate the earliest release date for rows of sequence entries given to it one by one.
+ * Assumes that rows are sorted: all accession entries are given in a block, and with ascending versions.
+ *
+ * The earliest release date of a sequence is the earliest date of:
+ * - the internal release date
+ * - any date from a given list of fields
+ * - the earliest release date from the previous version (if it exists)
+ */
+class EarliestReleaseDateFinder(private val fields: List<String>) {
+    private val earliestReleaseDateCache = mutableMapOf<String, LocalDateTime>()
+
+    fun calculateEarliestReleaseDate(rawProcessedData: RawProcessedData): LocalDateTime {
+        var earliestReleaseDate = rawProcessedData.releasedAtTimestamp
+
+        fields.forEach { field ->
+            rawProcessedData.processedData.metadata[field]?.textValue()?.let { dateText ->
+                val date = try {
+                    LocalDateTime(LocalDate.parse(dateText), LocalTime.fromSecondOfDay(0))
+                } catch (e: IllegalArgumentException) {
+                    log.warn {
+                        "Incorrectly formatted date on ${rawProcessedData.accession}." +
+                            "${rawProcessedData.version} on field $field: $dateText"
+                    }
```

Comment:
Yeah indeed this should never happen because inputs are already validated (as you pointed out earlier).

> If the latter, I think that we should indicate it more clearly and e.g. encourage the reader to report it.

Where would they report it? Could you maybe give me the message you'd like to see here?

Reply:
I'd use at least log level `error` – maybe something even higher.

What about that? "Unexpected error: Incorrectly formatted date on ${rawProcessedData.accession}. ${rawProcessedData.version} on field $field: $dateText – Something is wrong with this instance: it might be a configuration error or a bug of the software. Please feel free to reach out to the developers if you need advise (https://github.com/loculus-project/loculus/issues)"